File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-0.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public List<int> InorderTraversal(TreeNode root) {
        List<int> result = new List<int>();
        
        return recurse(root, result);

        
    }

    public List<int> recurse(TreeNode root, List<int> result) {
        if (root == null) {
            return result;
        }

        recurse(root.left, result);
        result.Add(root.val);
        recurse(root.right, result);

        return result;
    }
}

File: Data Structures & Algorithms/path-sum/submission-0.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
impl Solution {
    pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, cur_sum: i32, target: i32) -> bool {
            if let Some(n) = node {
                let n = n.borrow();
                let new_sum = cur_sum + n.val;
                if n.left.is_none() && n.right.is_none() {
                    return new_sum == target;
                }
                dfs(&n.left, new_sum, target) || dfs(&n.right, new_sum, target)
            } else {
                false
            }
        }
        dfs(&root, 0, target_sum)
    }
}
